engine/access/rest/websockets/data_providers/events_provider.go

```diff
@@ -0,0 +1,209 @@
+	topic string,
+	arguments models.Arguments,
+	send chan<- interface{},
+	chain flow.Chain,
+	eventFilterConfig state_stream.EventFilterConfig,
+	heartbeatInterval uint64,
+) (*EventsDataProvider, error) {
+	p := &EventsDataProvider{
+		logger:            logger.With().Str("component", "events-data-provider").Logger(),
+		stateStreamApi:    stateStreamApi,
+		heartbeatInterval: heartbeatInterval,
+	}
+
+	// Initialize arguments passed to the provider.
+	eventArgs, err := ParseEventsArguments(arguments, chain, eventFilterConfig)
+	if err != nil {
+		return nil, fmt.Errorf("invalid arguments for events data provider: %w", err)
+	}
+
+	subCtx, cancel := context.WithCancel(ctx)
+
+	p.baseDataProvider = newBaseDataProvider(
+		topic,
+		cancel,
+		send,
+		p.createSubscription(subCtx, eventArgs), // Set up a subscription to events based on arguments.
+	)
+
+	return p, nil
+}
+
+// Run starts processing the subscription for events and handles responses.
+//
+// No errors are expected during normal operations.
+func (p *EventsDataProvider) Run() error {
+	return subscription.HandleSubscription(p.subscription, p.handleResponse(p.send))
+}
+
+func (p *EventsDataProvider) handleResponse(send chan<- interface{}) func(eventsResponse *backend.EventsResponse) error {
+	blocksSinceLastMessage := uint64(0)
+	messageIndex := counters.NewMonotonousCounter(1)
+
+	return func(eventsResponse *backend.EventsResponse) error {
+		// check if there are any events in the response. if not, do not send a message unless the last
+		// response was more than HeartbeatInterval blocks ago
+		if len(eventsResponse.Events) == 0 {
+			blocksSinceLastMessage++
+			if blocksSinceLastMessage < p.heartbeatInterval {
+				return nil
+			}
+			blocksSinceLastMessage = 0
+		}
+
+		index := messageIndex.Value()
+		if ok := messageIndex.Set(messageIndex.Value() + 1); !ok {
+			return status.Errorf(codes.Internal, "message index already incremented to %d", messageIndex.Value())
+		}
+
+		send <- &models.EventResponse{
+			BlockId:        eventsResponse.BlockID.String(),
+			BlockHeight:    strconv.FormatUint(eventsResponse.Height, 10),
+			BlockTimestamp: eventsResponse.BlockTimestamp,
+			Events:         eventsResponse.Events,
+			MessageIndex:   strconv.FormatUint(index, 10),
+		}
+
+		return nil
+	}
+}
+
+// createSubscription creates a new subscription using the specified input arguments.
+func (p *EventsDataProvider) createSubscription(ctx context.Context, args EventsArguments) subscription.Subscription {
+	if args.StartBlockID != flow.ZeroID {
+		return p.stateStreamApi.SubscribeEventsFromStartBlockID(ctx, args.StartBlockID, args.Filter)
+	}
+
+	if args.StartBlockHeight != request.EmptyHeight {
+		return p.stateStreamApi.SubscribeEventsFromStartHeight(ctx, args.StartBlockHeight, args.Filter)
+	}
+
+	return p.stateStreamApi.SubscribeEventsFromLatest(ctx, args.Filter)
+}
+
+// ParseEventsArguments validates and initializes the events arguments.
+func ParseEventsArguments(
```

Comment:
Will `ParseEventsArguments` be used somewhere else than in `EventsDataProvider`?  If not, I believe it should be part of `EventsDataProvider` and not public